samcli/lib/providers/sam_layer_provider.py

```diff
@@ -138,11 +149,21 @@ def _convert_lambda_layer_resource(
             LOG.debug("--base-dir is not presented, adjusting uri %s relative to %s", codeuri, stack.location)
             codeuri = SamLocalStackProvider.normalize_resource_path(stack.location, codeuri)
 
+        # If the layer from CDK, name will be CDK resource id.
+        # Normally, LayerVersion will infer the name and layer id from the arn,
+        # but if the logical id is different from the name,
+        # specify the name and layer id in LayerVersion to use that name.
+        name = ResourceMetadataNormalizer.get_resource_name(resource_properties, layer_logical_id)
+        name = name if name != layer_logical_id else None
```

Comment:
I think I will need more time to digest this. But it's not super clear why this is here, how this changes part of the code base for getting the name or logical id for this resource.

Reply:
The code was indeed a bit difficult to read the intent. This is a change that applies only to CDK resources, and requires that `layer_id` and `name` be specified when initializing the `LayerVersion` as described at https://github.com/aws/aws-sam-cli/pull/4064#discussion_r939480604 This is a change that applies only to CDK resources.
I have refactored it so that it is clear that it applies only to CDK resources.

Reply:
Maybe layer is different but it might be easier to understand and read if we don't do conditional things per IAC. This can make expanding and supporting Terraform or other IAC much more difficult in the future.